drop-transfer/src/ws/client/socket.rs

```diff
@@ -0,0 +1,56 @@
+use std::{
+    io,
+    time::{Duration, Instant},
+};
+
+use futures_util::{SinkExt, StreamExt};
+use tokio::net::TcpStream;
+use tokio_tungstenite::{tungstenite::Message, WebSocketStream};
+
+pub type WsStream = WebSocketStream<TcpStream>;
+
+pub struct WebSocket {
+    stream: WsStream,
+    recv_last: Instant,
+    recv_timeout: Duration,
+    send_timeout: Duration,
+}
+
+impl WebSocket {
+    pub fn new(stream: WsStream, recv_timeout: Duration, send_timeout: Duration) -> Self {
+        Self {
+            stream,
+            recv_last: Instant::now(),
+            recv_timeout,
+            send_timeout,
+        }
+    }
+
+    pub async fn send(&mut self, msg: Message) -> crate::Result<()> {
+        tokio::time::timeout(self.send_timeout, self.stream.send(msg))
+            .await
+            .map_err(|err| io::Error::new(io::ErrorKind::TimedOut, err))??;
```

Comment:
I am curious why do we need 2x`?`? Can `map_err` fail to convert the message..?

Reply:
No, there are two levels of error here

Reply:
Oh okay `stream.send` might error as well